---
title: "Attack of the clones initial locus filtering"
author: "Peter Euclide"
date: "May 23, 2019"
output: html_document
---

***


```{r setup, include=FALSE}

library(tidyr)
library(dplyr)
library(ggplot2)
library(ggpubr)

```


```{r read in allele count data, include=FALSE}
### Read in species specific parsed data files. Files were generated by parsing VCF fles that can be found on Dryrad using custom a Python script.


## Raw data of reads per allele, these file names change for each species

CF_reads_per_allele_RAW <- read.delim("./CF_reads_per_allele_vcf_parsed.txt", header=TRUE) 
NO_CF_reads_per_allele_RAW = read.delim("./NO_CF_reads_per_allele_vcf_parsed.txt", header=TRUE)

```


```{r read in genotype calls data, include=FALSE}

## Raw genotype calls, these file names change for each species

CF_parsed_genotype_calls <- read.delim("./CF_vcf_parsed_genotypes.txt", header=TRUE)
NO_CF_parsed_genotype_calls <- read.delim("./NO_CF_vcf_parsed_genotypes.txt", header = TRUE)


```


```{r read in read depth data, include=FALSE}

## Raw read depth, these file names change for each species

CF_parsed_read_depth <- read.delim("./CF_read_depth_vcf_parsed.txt", header=TRUE)
NO_CF_parsed_read_depth <- read.delim("./NO_CF_read_depth_vcf_parsed.txt", header=TRUE)


```


```{r filter PSVs}

## HDPLot (McKinney, Waples, Seeb, & Seeb, 2017) was used to identify the best SNPs based on allele count and heterozygosity and datasets were filtered to retain only these loci.

no_PSV_data <- read.table("./HD_plot/HDplotResults.txt", header = T)

# create black list
psv_data <- CF_reads_per_allele_RAW[!CF_reads_per_allele_RAW$ID %in% no_PSV_data$Locus_ID,1:2]
# UD loci
psv_data <- psv_data %>% separate("ID", into =c("locus", "pos"))
psv_loci <- unique(psv_data$locus)

## now we are removing all loci that have at least one putative PSV - update 1-14-19

# reads per allele

CF_reads_per_allele_RAW <- CF_reads_per_allele_RAW %>% separate("ID", into = c("locus", "pos"), remove = F)%>% filter(!locus %in% psv_loci)
CF_reads_per_allele_RAW <- CF_reads_per_allele_RAW[,-c(2:3)] 


NO_CF_reads_per_allele_RAW <- NO_CF_reads_per_allele_RAW %>% separate("ID", into = c("locus", "pos"), remove = F)%>% filter(!locus %in% psv_loci)
NO_CF_reads_per_allele_RAW <- NO_CF_reads_per_allele_RAW[,-c(2:3)] 

# geneotype call
CF_parsed_genotype_calls <- CF_parsed_genotype_calls %>% separate("ID", into = c("locus", "pos"), remove = F)%>% filter(!locus %in% psv_loci)
CF_parsed_genotype_calls <- CF_parsed_genotype_calls[,-c(2:3)] 

NO_CF_parsed_genotype_calls <- NO_CF_parsed_genotype_calls %>% separate("ID", into = c("locus", "pos"), remove = F)%>% filter(!locus %in% psv_loci)
NO_CF_parsed_genotype_calls <- NO_CF_parsed_genotype_calls[,-c(2:3)] 



# read depth
CF_parsed_read_depth <- CF_parsed_read_depth %>% separate("ID", into = c("locus", "pos"), remove = F)%>% filter(!locus %in% psv_loci)
CF_parsed_read_depth <- CF_parsed_read_depth[,-c(2:3)] 

NO_CF_parsed_read_depth <- NO_CF_parsed_read_depth %>% separate("ID", into = c("locus", "pos"), remove = F)%>% filter(!locus %in% psv_loci)
NO_CF_parsed_read_depth <- NO_CF_parsed_read_depth[,-c(2:3)]

```


```{r find missing samples between matricies, include=FALSE}
# change CF names to be same as non-cf names
CF_names <- colnames(CF_parsed_genotype_calls[,c(7:length(CF_parsed_genotype_calls))])
new_names <- substr(CF_names,1,nchar(CF_names)-2)

names(CF_parsed_genotype_calls)[c(7:length(CF_parsed_genotype_calls))] <- new_names
names(CF_parsed_read_depth)[c(7:length(CF_parsed_read_depth))] <- new_names
names(CF_reads_per_allele_RAW)[c(7:length(CF_reads_per_allele_RAW))] <- new_names



## subset dataset to only loci and samples shared between both the CF and non-CF datasets
missingNames <- names(CF_reads_per_allele_RAW)[!names(CF_reads_per_allele_RAW) %in% names(NO_CF_reads_per_allele_RAW)]
commonNames <- names(CF_reads_per_allele_RAW)[names(CF_reads_per_allele_RAW) %in% names(NO_CF_reads_per_allele_RAW)]

# filter both CF and No_CF dataset to only common samples

# reads per allele
CF_reads_per_allele_RAW_filtered <- CF_reads_per_allele_RAW[,commonNames]
NO_CF_reads_per_allele_RAW_filtered <- NO_CF_reads_per_allele_RAW[,commonNames]

# genotype calls
CF_parsed_genotype_calls_filtered <- CF_parsed_genotype_calls[,commonNames]
NO_CF_parsed_genotype_calls_filtered <- NO_CF_parsed_genotype_calls[,commonNames]

# read depth
CF_parsed_read_depth_filtered <- CF_parsed_read_depth[,commonNames]
NO_CF_parsed_read_depth_filtered <- NO_CF_parsed_read_depth[,commonNames]

```


```{r filter to common set of loci, include=FALSE}

#ID common loci
commonLoci <- intersect(NO_CF_reads_per_allele_RAW_filtered$ID, CF_reads_per_allele_RAW_filtered$ID)
length(commonLoci)
# reads per allele
CF_reads_per_allele_RAW_filtered <- droplevels(subset(CF_reads_per_allele_RAW_filtered, ID %in% commonLoci))
NO_CF_reads_per_allele_RAW_filtered <- droplevels(subset(NO_CF_reads_per_allele_RAW_filtered, ID %in% commonLoci))

# genotype calls
CF_parsed_genotype_calls_filtered <- droplevels(subset(CF_parsed_genotype_calls_filtered, ID %in% commonLoci))
NO_CF_parsed_genotype_calls_filtered <- droplevels(subset(NO_CF_parsed_genotype_calls_filtered, ID %in% commonLoci))

# read depth
CF_parsed_read_depth_filtered <- droplevels(subset(CF_parsed_read_depth_filtered, ID %in% commonLoci))
NO_CF_parsed_read_depth_filtered <- droplevels(subset(NO_CF_parsed_read_depth_filtered, ID %in% commonLoci))

```


```{r wide -> long format, include=FALSE}
## Convert dataframes to long format for easier manipulation

# reads per allele
CF_reads_per_allele_RAW_filtered_long <- CF_reads_per_allele_RAW_filtered %>% 
    gather(sample, reads, 7:length(CF_reads_per_allele_RAW_filtered))
NO_CF_reads_per_allele_RAW_filtered_long <- NO_CF_reads_per_allele_RAW_filtered %>% 
    gather(sample, reads, 7:length(NO_CF_reads_per_allele_RAW_filtered))

# genotype calls
CF_parsed_genotype_calls_filtered_long <- CF_parsed_genotype_calls_filtered %>% 
    gather(sample, call, 7:length(CF_parsed_genotype_calls_filtered))
NO_CF_parsed_genotype_calls_filtered_long <- NO_CF_parsed_genotype_calls_filtered %>% 
    gather(sample, call, 7:length(NO_CF_parsed_genotype_calls_filtered))

# read depth
CF_parsed_read_depth_filtered_long <- CF_parsed_read_depth_filtered %>% 
    gather(sample, depth, 7:length(CF_parsed_read_depth_filtered))
NO_CF_parsed_read_depth_filtered_long <- NO_CF_parsed_read_depth_filtered %>% 
    gather(sample, depth, 7:length(NO_CF_parsed_read_depth_filtered))

```


```{r filter to good samples and loci, include=FALSE}
## Filter data to only individuals with less than 70 % missing data and Loci present in > 50 % of inds

# find number of loci
nLoci <- length(unique(CF_parsed_genotype_calls_filtered_long$ID))

#summarize to find inds that have less than 70 % missing data CF vs. NO CF
Ind_quality <- CF_parsed_genotype_calls_filtered_long %>%
  group_by(sample) %>%
    summarise_at(vars(call),funs(n(), (sum(. == -9)/nLoci < .7)) ) ### I THINK THIS SHOULD BE LESS THAN 30% NOT 70%...

goodInds <- Ind_quality[Ind_quality[,3] == T,"sample"]
nrow(goodInds)


#no CF

Ind_quality_NO_CF <- NO_CF_parsed_genotype_calls_filtered_long %>%
  group_by(sample) %>%
    summarise_at(vars(call),funs(n(), (sum(. == -9)/nLoci < .7)) )

goodInds_NO_CF <- Ind_quality[Ind_quality[,3] == T,"sample"]
nrow(goodInds_NO_CF)

paste0("Individuals lost due to quality NO_CF = ", length(unique(NO_CF_parsed_genotype_calls_filtered_long$sample))-nrow(goodInds_NO_CF))

paste0("Individuals lost due to quality CF = ", length(unique(NO_CF_parsed_genotype_calls_filtered_long$sample))-nrow(goodInds))

# filter datasets by good individuals in CF dataset ---

# reads per allele
CF_reads_per_allele_RAW_filtered_long <- droplevels(subset(CF_reads_per_allele_RAW_filtered_long, sample %in% goodInds$sample))
NO_CF_reads_per_allele_RAW_filtered_long <- droplevels(subset(NO_CF_reads_per_allele_RAW_filtered_long, sample %in% goodInds$sample))

# genotype calls
CF_parsed_genotype_calls_filtered_long <- droplevels(subset(CF_parsed_genotype_calls_filtered_long, sample %in% goodInds$sample))
NO_CF_parsed_genotype_calls_filtered_long <- droplevels(subset(NO_CF_parsed_genotype_calls_filtered_long, sample %in% goodInds$sample))

# read depth
CF_parsed_read_depth_filtered_long <- droplevels(subset(CF_parsed_read_depth_filtered_long, sample %in% goodInds$sample))
NO_CF_parsed_read_depth_filtered_long <- droplevels(subset(NO_CF_parsed_read_depth_filtered_long, sample %in% goodInds$sample))


# now find only quality loci ----

nInds <- length(goodInds$sample)  



Loc_quality <-CF_parsed_genotype_calls_filtered_long %>% 
  group_by(ID) %>% 
  summarise(GT_rate = sum(call != -9)/length(ID),
            sumCall= sum(call != -9),
            lengthID =length(ID), 
            quality = GT_rate > .5)


goodLoci_cf <- Loc_quality %>% filter(quality == TRUE)
badLoci <- Loc_quality %>% filter(quality == FALSE)

nrow(goodLoci_cf)
#sum(Loc_quality$`(`)


######## 


Loc_quality <-NO_CF_parsed_genotype_calls_filtered_long %>% 
  group_by(ID) %>% 
  summarise(GT_rate = sum(call != -9)/length(ID),
            sumCall= sum(call != -9),
            lengthID =length(ID), 
            quality = GT_rate > .5)
goodLoci_nocf <- Loc_quality %>% filter(quality == TRUE)



## good loci in both datasets

goodLoci <- intersect(goodLoci_nocf$ID, goodLoci_cf$ID)


# filter datasets by good loci---

# reads per allele

CF_reads_per_allele_RAW_filtered_long <- CF_reads_per_allele_RAW_filtered_long %>% filter(ID %in% goodLoci)

NO_CF_reads_per_allele_RAW_filtered_long <- NO_CF_reads_per_allele_RAW_filtered_long%>% filter(ID %in% goodLoci)

# genotype calls
CF_parsed_genotype_calls_filtered_long <- CF_parsed_genotype_calls_filtered_long%>% filter(ID %in% goodLoci)
NO_CF_parsed_genotype_calls_filtered_long <- NO_CF_parsed_genotype_calls_filtered_long%>% filter(ID %in% goodLoci)
# read depth
CF_parsed_read_depth_filtered_long <- CF_parsed_read_depth_filtered_long%>% filter(ID %in% goodLoci)
NO_CF_parsed_read_depth_filtered_long <- NO_CF_parsed_read_depth_filtered_long%>% filter(ID %in% goodLoci)



```


```{r reference alleles, include=FALSE}

## ID flipped reference alleles

commonRefs <- NO_CF_reads_per_allele_RAW_filtered[NO_CF_reads_per_allele_RAW_filtered$REF1== CF_reads_per_allele_RAW_filtered$REF1, "ID"]
flippedRefs <- NO_CF_reads_per_allele_RAW_filtered[!NO_CF_reads_per_allele_RAW_filtered$ID %in% commonRefs, "ID"]


## number of flipped Refs
paste0("Number of flipped references = ", length(flippedRefs))


```


```{r convert reference alleles, include=FALSE}

## convert flipped reference alleles

FRefs_CF_reads_per_allele_RAW_filtered_long <- CF_reads_per_allele_RAW_filtered_long[CF_reads_per_allele_RAW_filtered_long$ID %in% flippedRefs,]
FRefs_CF_reads_per_allele_RAW_filtered_long$flipped <- paste(sub(".*,", "", FRefs_CF_reads_per_allele_RAW_filtered_long$reads), 
                                                             sub(",.*", "", FRefs_CF_reads_per_allele_RAW_filtered_long$reads), sep = ",")

FRefs_CF_parsed_genotype_calls_filtered_long <- CF_parsed_genotype_calls_filtered_long[CF_parsed_genotype_calls_filtered_long$ID %in% flippedRefs,]
#change calls so that 2 = 0 and 0 = 2
flipped <- NULL
#FRefs_CF_parsed_genotype_calls_filtered_long$flipped <- 
  
FRefs_CF_parsed_genotype_calls_filtered_long$flipped <- ifelse(FRefs_CF_parsed_genotype_calls_filtered_long$call == 2, 0, FRefs_CF_parsed_genotype_calls_filtered_long$call)
FRefs_CF_parsed_genotype_calls_filtered_long$flipped <- ifelse(FRefs_CF_parsed_genotype_calls_filtered_long$call == 0, 2, FRefs_CF_parsed_genotype_calls_filtered_long$flipped)
FRefs_CF_parsed_genotype_calls_filtered_long$flipped <- ifelse(FRefs_CF_parsed_genotype_calls_filtered_long$call == 1, 1, FRefs_CF_parsed_genotype_calls_filtered_long$flipped)


FRefs_CF_parsed_read_depth_filtered_long <- CF_parsed_read_depth_filtered_long[CF_parsed_read_depth_filtered_long$ID %in% flippedRefs,]


```

```{r, include=FALSE}
##change calls to be flipped for messed up loci or filter them out.


## for now just filter these out and see how that effects things:

# # reads per allele
# CF_reads_per_allele_RAW_filtered_long <- droplevels(subset(CF_reads_per_allele_RAW_filtered_long, ID %in% commonRefs))
# NO_CF_reads_per_allele_RAW_filtered_long <- droplevels(subset(NO_CF_reads_per_allele_RAW_filtered_long, ID %in% commonRefs))
# 
# # genotype calls
# CF_parsed_genotype_calls_filtered_long <- droplevels(subset(CF_parsed_genotype_calls_filtered_long, ID %in% commonRefs))
# NO_CF_parsed_genotype_calls_filtered_long <- droplevels(subset(NO_CF_parsed_genotype_calls_filtered_long, ID %in% commonRefs))
# 
# # read depth
# CF_parsed_read_depth_filtered_long <- droplevels(subset(CF_parsed_read_depth_filtered_long, ID %in% commonRefs))
# NO_CF_parsed_read_depth_filtered_long <- droplevels(subset(NO_CF_parsed_read_depth_filtered_long, ID %in% commonRefs))


### OR you can change calls to flipped calls - this is what we did for the manuscript 

CF_reads_per_allele_RAW_filtered_long[CF_reads_per_allele_RAW_filtered_long$ID %in% flippedRefs, "reads"]  <-  FRefs_CF_reads_per_allele_RAW_filtered_long$flipped
CF_parsed_genotype_calls_filtered_long[CF_parsed_genotype_calls_filtered_long$ID %in% flippedRefs,"call"] <- FRefs_CF_parsed_genotype_calls_filtered_long$flipped

```


```{r label libraries, include=FALSE}

## Because library information was not included in VCF files, we used the original barcodes files to call individuals by name and add library metadata

lib_info <- read.table("./library_info.txt", header = T)
lib_info$sample <- gsub("-", ".", lib_info$sample, fixed = T)
# add lib column to data
# reads per allele
CF_reads_per_allele_RAW_filtered_long <- merge(CF_reads_per_allele_RAW_filtered_long, lib_info, by = "sample")
NO_CF_reads_per_allele_RAW_filtered_long <- merge(NO_CF_reads_per_allele_RAW_filtered_long, lib_info, by = "sample")

# genotype calls
CF_parsed_genotype_calls_filtered_long <- merge(CF_parsed_genotype_calls_filtered_long, lib_info, by = "sample")
NO_CF_parsed_genotype_calls_filtered_long <- merge(NO_CF_parsed_genotype_calls_filtered_long, lib_info, by = "sample")

# read depth
CF_parsed_read_depth_filtered_long <- merge(CF_parsed_read_depth_filtered_long, lib_info, by = "sample")
NO_CF_parsed_read_depth_filtered_long <- merge(NO_CF_parsed_read_depth_filtered_long, lib_info, by = "sample")



```

```{r write filtered files}
## the final step was to write filtered and unfiltered datasets for each species to text files for use in analysis.


write.table(NO_CF_parsed_genotype_calls_filtered_long, "./walleye/NO_CF_wa_parsed_genotype_calls_filtered_long_1-14-19")
write.table(CF_parsed_genotype_calls_filtered_long, "./walleye/CF_wa_parsed_genotype_calls_filtered_long_1-14-19")

write.table(NO_CF_parsed_read_depth_filtered_long, "./walleye/NO_CF_wa_parsed_read_depth_filtered_long_1-14-19")
write.table(CF_parsed_read_depth_filtered_long, "./walleye/CF_wa_parsed_read_depth_filtered_long_1-14-19")

write.table(NO_CF_reads_per_allele_RAW_filtered_long, "./walleye/NO_CF_wa_reads_per_allele_RAW_filtered_long_1-14-19")
write.table(CF_reads_per_allele_RAW_filtered_long, "./walleye/CF_wa_reads_per_allele_RAW_filtered_long_1-14-19")

```
